### PR TITLE
Add the function of supporting static pod on edge nodes.

### PR DESCRIPTION
**Pull request by luomengY**
What type of PR is this?

/kind feature

What this PR does / why we need it:

Support static pod at edgenodes

### CODE REVIEW

**Comment by luomengY** [comment]
This is the test effect.@Shelley-BaoYue
be6d54e67988e6f2c92e98bd871e148
9bbc2c71a088a41e51434edb1adc940

**Review by fisherxu — COMMENTED** [review]
@luomengY Thanks for the contributions, I'll take a look :)

File: cloud/pkg/cloudhub/dispatcher/message_dispatcher.go [hunk 1]
```diff
@@ -467,7 +467,8 @@ func noAckRequired(msg *beehivemodel.Message) bool {
 				resourceType == beehivemodel.ResourceTypeLease ||
 				resourceType == beehivemodel.ResourceTypeNodePatch ||
 				resourceType == beehivemodel.ResourceTypePodPatch ||
-				resourceType == beehivemodel.ResourceTypePodStatus {
+				resourceType == beehivemodel.ResourceTypePodStatus ||
+				resourceType == beehivemodel.ReSourceTypeCreateMirrorPod {
```
Comment (Shelley-BaoYue) [inline cloud/pkg/cloudhub/dispatcher/message_dispatcher.go:471]: We can make it more generic by directly using ReSourceTypeCreatePod

Comment (Shelley-BaoYue) [reply cloud/pkg/cloudhub/dispatcher/message_dispatcher.go:471]: RessourceTypeCreatePod

File: cloud/pkg/edgecontroller/controller/upstream.go [hunk 2]
```diff
@@ -116,6 +116,7 @@ type UpstreamController struct {
 	ruleStatusChan            chan model.Message
 	createLeaseChan           chan model.Message
 	queryLeaseChan            chan model.Message
+	createMirrorPodChan       chan model.Message
```
Comment (Shelley-BaoYue) [inline cloud/pkg/edgecontroller/controller/upstream.go:119]: ditto. using createPodChan

File: edge/pkg/metamanager/process.go [hunk 3]
```diff
@@ -401,6 +409,8 @@ func (m *metaManager) process(message model.Message) {
 		m.processUpdate(message)
 	case model.PatchOperation:
 		m.processPatch(message)
+	case model.CreateMirrorPod:
```
Comment (Shelley-BaoYue) [inline edge/pkg/metamanager/process.go:412]: directly using insert operation is ok

File: keadm/cmd/keadm/app/cmd/edge/image.go [hunk 4]
```diff
@@ -18,12 +18,20 @@ package edge
 
 import (
 	"fmt"
+	"github.com/kubeedge/kubeedge/common/constants"
```
Comment (Shelley-BaoYue) [inline keadm/cmd/keadm/app/cmd/edge/image.go:21]: please sort the import.

**Comment by Shelley-BaoYue** [comment]
/hold  
This PR depends on kubeedge/kubernetes updates.

**Comment by luomengY** [comment]
The static pod function code for the kubeedge/kubernetes branch has been submitted.




---Original---
From: "Shelley ***@***.***&gt;
Date: Wed, Aug 23, 2023 10:05 AM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [kubeedge/kubeedge] Resolve the issue of system bin directorybeing deleted by edgecore after mqtt crashes and crashes at edge nodes (PR#4825)




 
/hold
 This PR depends on kubeedge/kubernetes updates.

**Comment by luomengY** [comment]
> /hold This PR depends on kubeedge/kubernetes updates.

The static pod function code for the kubeedge/kubernetes branch has been submitted.

File: cloud/pkg/edgecontroller/controller/upstream.go [hunk 5]
```diff
@@ -1042,6 +1049,54 @@ func (uc *UpstreamController) patchPod() {
...
+				klog.Warningf("message: %s process failure, get namespace failed with error: %v", msg.GetID(), err)
+				continue
+			}
+			name, err := messagelayer.GetResourceName(msg)
+			if err != nil {
+				klog.Warningf("message: %s process failure, get resource name failed with error: %v", msg.GetID(), err)
+				continue
+			}
+
+			mirrorPodBytes, err := msg.GetContentData()
```
Comment (Shelley-BaoYue) [inline cloud/pkg/edgecontroller/controller/upstream.go:1071]: mirrorPodBytes --> podBytes

File: cloud/pkg/edgecontroller/controller/upstream.go [hunk 6]
```diff
@@ -1042,6 +1049,54 @@ func (uc *UpstreamController) patchPod() {
...
+			if err != nil {
+				klog.Warningf("message: %s process failure, get data failed with error: %v", msg.GetID(), err)
+				continue
+			}
+			var pod v1.Pod
+			if err = json.Unmarshal(mirrorPodBytes, &pod); err != nil {
+				klog.Errorf("mirrorPod  unmarshal faild")
```
Comment (Shelley-BaoYue) [inline cloud/pkg/edgecontroller/controller/upstream.go:1078]: "unmarshal pod request failed" and log the err

File: cloud/pkg/edgecontroller/controller/upstream.go [hunk 7]
```diff
@@ -1042,6 +1049,54 @@ func (uc *UpstreamController) patchPod() {
...
+				continue
+			}
+
+			createPod, err := uc.kubeClient.CoreV1().Pods(namespace).Create(context.TODO(), &pod, metaV1.CreateOptions{})
+			if err != nil {
+				klog.Errorf("message: %s process failure, patch pod failed with error: %v, namespace: %s, name: %s", msg.GetID(), err, namespace, name)
```
Comment (Shelley-BaoYue) [inline cloud/pkg/edgecontroller/controller/upstream.go:1084]: patch pod --> create pod

File: cloud/pkg/edgecontroller/controller/upstream.go [hunk 8]
```diff
@@ -1042,6 +1049,54 @@ func (uc *UpstreamController) patchPod() {
...
+
+			resMsg := model.NewMessage(msg.GetID()).
+				SetResourceVersion(createPod.ResourceVersion).
+				FillBody(&edgeapi.ObjectResp{Object: createPod, Err: err}).
+				BuildRouter(modules.EdgeControllerModuleName, constants.GroupResource, msg.GetResource(), model.ResponseOperation)
+			if err = uc.messageLayer.Response(*resMsg); err != nil {
+				klog.Errorf("Message: %s process failure, response failed with error: %v", msg.GetID(), err)
+				continue
+			}
+			klog.V(4).Infof("message: %s, create mirrorPod successfully, namespace: %s, name: %s", msg.GetID(), namespace, name)
```
Comment (Shelley-BaoYue) [inline cloud/pkg/edgecontroller/controller/upstream.go:1095]: create pod

File: common/constants/default.go [hunk 9]
```diff
@@ -52,14 +52,15 @@ const (
...
-	DefaultMosquittoImage        = "eclipse-mosquitto:1.6.15"
+	DefaultRootDir                = "/var/lib/edged"
+	DefaultDockerAddress          = "unix:///var/run/docker.sock"
+	DefaultRuntimeType            = "remote"
+	DefaultDockershimRootDir      = "/var/lib/dockershim"
+	DefaultEdgedMemoryCapacity    = 7852396000
+	DefaultRemoteRuntimeEndpoint  = "unix:///run/containerd/containerd.sock"
+	DefaultRemoteImageEndpoint    = "unix:///run/containerd/containerd.sock"
+	DefaultMosquittoImage         = "eclipse-mosquitto:1.6.15"
+	DeafultMosquittoContianerName = "mqtt"
```
Comment (Shelley-BaoYue) [inline common/constants/default.go:63]: DeafultMosquittoContianerName can be removed

File: edge/pkg/metamanager/client/pod.go [hunk 10]
```diff
@@ -52,7 +52,19 @@ func newPods(namespace string, s SendInterface) *pods {
 }
 
 func (c *pods) Create(cm *corev1.Pod) (*corev1.Pod, error) {
-	return nil, nil
+	resource := fmt.Sprintf("%s/%s/%s", c.namespace, model.ReSourceTypeCreatePod, cm.Name)
+	podMsg := message.BuildMsg(modules.MetaGroup, "", modules.EdgedModuleName, resource, model.InsertOperation, *cm)
+	resp, err := c.send.SendSync(podMsg)
+	if err != nil {
+		return nil, fmt.Errorf("update pod failed, err: %v", err)
```
Comment (Shelley-BaoYue) [inline edge/pkg/metamanager/client/pod.go:59]: create pod

File: keadm/cmd/keadm/app/cmd/common/constant.go [hunk 11]
```diff
@@ -119,6 +119,9 @@ const (
 	// EdgecoreConfigPath is default edgecore config path
 	EdgecoreConfigPath = "/etc/kubeedge/config/edgecore.yaml"
 
+	// DefaultManifestsDir edge node default static pod path
```
Comment (Shelley-BaoYue) [inline keadm/cmd/keadm/app/cmd/common/constant.go:122]: Is it better to put this contant in /common/contants/default.go?

File: staging/src/github.com/kubeedge/beehive/pkg/core/model/message.go [hunk 12]
```diff
@@ -36,6 +36,7 @@ const (
 	ResourceTypeRuleStatus          = "rulestatus"
 	ResourceTypeLease               = "lease"
 	ResourceTypeSaAccess            = "serviceaccountaccess"
+	ReSourceTypeCreatePod           = "createpod"
```
Comment (Shelley-BaoYue) [inline staging/src/github.com/kubeedge/beehive/pkg/core/model/message.go:39]:
```suggestion
	ResourceTypeCreatePod           = "createpod"
```

File: pkg/apis/componentconfig/cloudcore/v1alpha1/types.go [hunk 13]
```diff
@@ -419,7 +424,7 @@ type DynamicController struct {
 	Enable bool `json:"enable"`
 }
 
-// CloudStream indicates the stream controller
+// CloudSream indicates the stream controller
```
Comment (fisherxu) [inline pkg/apis/componentconfig/cloudcore/v1alpha1/types.go:427]:
```suggestion
// CloudStream indicates the stream controller
```

**Review by fisherxu — COMMENTED** [review]
Overall looks good, It's need to make CI happy :)

**Review by fisherxu — COMMENTED** [review]
/approve
/assign @Shelley-BaoYue

**Review by fisherxu — COMMENTED** [review]
/hold cancel 
/lgtm

**Pull request merged**